tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,344 @@
+  const auto& deriv_phi = get<
+      Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>, Frame::Inertial>>(
+      sw_derivs);
+
+  // Insert into databox
+  const auto box = db::create<
+      db::AddSimpleTags<
+          ::Tags::Coordinates<dim, Frame::Inertial>, ScalarWave::Psi,
+          ScalarWave::Phi<dim>, ScalarWave::Pi,
+          ::Tags::deriv<ScalarWave::Psi, tmpl::size_t<dim>, Frame::Inertial>,
+          ::Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>,
+                        Frame::Inertial>,
+          ::Tags::deriv<ScalarWave::Pi, tmpl::size_t<dim>, Frame::Inertial>>,
+      db::AddComputeTags<ScalarWave::Tags::OneIndexConstraintCompute<dim>,
+                         ScalarWave::Tags::TwoIndexConstraintCompute<dim>>>(
+      x, psi, phi, pi, deriv_psi, deriv_phi, deriv_pi);
+
+  // Compute tested quantities locally
+  const auto one_index_constraint =
+      ScalarWave::one_index_constraint(deriv_psi, phi);
+  const auto two_index_constraint = ScalarWave::two_index_constraint(deriv_phi);
+
+  // Check that their compute items in databox furnish identical values
+  CHECK(db::get<ScalarWave::Tags::OneIndexConstraint<dim>>(box) ==
+        one_index_constraint);
+  CHECK(db::get<ScalarWave::Tags::TwoIndexConstraint<dim>>(box) ==
+        two_index_constraint);
+}
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.Constraints",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/ScalarWave/"};
+  INFO("Testing constraints with randomized input");
+  // Test the one-index constraint with random numbers
+  test_one_index_constraint_random<1>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+  test_one_index_constraint_random<2>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+  test_one_index_constraint_random<3>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+
+  // Test the two-index constraint with random numbers
+  test_two_index_constraint_random<1>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+  test_two_index_constraint_random<2>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+  test_two_index_constraint_random<3>(
+      DataVector(4, std::numeric_limits<double>::signaling_NaN()));
+
+  // Test constraints against Spherical wave
+  INFO("Testing constraints for a spherical-wave analytic solution");
+  const size_t grid_size = 8;
+  const std::array<double, 3> upper_bound{{6., 6., 6.}};
+  const std::array<double, 3> lower_bound{{0., 0., 0.}};
+
+  const ScalarWave::Solutions::RegularSphericalWave solution(
+      std::make_unique<MathFunctions::Gaussian>(1., 1., 0.));
+
+  // Note: looser numerical tolerance because this check
+  // uses numerical derivatives
+  test_one_index_constraint_analytic(
+      solution, grid_size, lower_bound, upper_bound,
+      std::numeric_limits<double>::epsilon() * 1.e6);
+
```

Comment:
Just give the tolerance explicitly eg as `1.e-10` since it should actually scale with the machine epsilon I think

Reply:
Isn't it better to have the tolerance scale with machine precision? Whats the harm anyway?

Reply:
I agree with @prayush. Better to do it relative to machine precision.

Reply:
Why? If I understand correctly, the precision is limited by discretization on the `grid_size`, not by the machine precision. Therefore, it has nothing to do really with the machine epsilon. If we ran on a machine with way higher numerical precision, we wouldn't get a more precise answer, so this test would fail. Am I getting something wrong?